fluent-syntax/src/parser/mod.rs

```diff
@@ -1,14 +1,842 @@
+            ps.ptr = line_start;
+            break;
+        }
+
+        match get_attribute(ps) {
+            Ok(attr) => attributes.push(attr),
+            Err(_) => {
+                ps.ptr = line_start;
+                break;
+            }
+        }
+    }
+    Ok(attributes)
+}
+
+fn get_attribute<'p>(ps: &mut ParserStream<'p>) -> Result<ast::Attribute<'p>> {
+    ps.expect_byte(b'.')?;
+    let id = get_identifier(ps)?;
+    ps.skip_blank_inline();
+    ps.expect_byte(b'=')?;
+    let pattern = get_pattern(ps)?;
+
+    match pattern {
+        Some(pattern) => Ok(ast::Attribute { id, value: pattern }),
+        None => error!(ErrorKind::MissingValue, ps.ptr),
+    }
+}
+
+fn get_identifier<'p>(ps: &mut ParserStream<'p>) -> Result<ast::Identifier<'p>> {
+    let mut ptr = ps.ptr;
+
+    while let Some(b) = ps.source.get(ptr) {
+        if ps.is_byte_alphabetic(*b) {
+            ptr += 1;
+        } else if ptr == ps.ptr {
+            return error!(
+                ErrorKind::ExpectedCharRange {
+                    range: "a-zA-Z".to_string()
+                },
+                ptr
+            );
+        } else if ps.is_byte_digit(*b) || [b'_', b'-'].contains(&b) {
+            ptr += 1;
+        } else {
+            break;
+        }
+    }
+    let name = ps.get_slice(ps.ptr, ptr);
+    ps.ptr = ptr;
+
+    Ok(ast::Identifier { name })
+}
+
+fn get_variant_key<'p>(ps: &mut ParserStream<'p>) -> Result<ast::VariantKey<'p>> {
+    if !ps.take_if(b'[') {
+        return error!(ErrorKind::ExpectedToken('['), ps.ptr);
+    }
+    ps.skip_blank();
+
+    let key = if ps.is_number_start() {
+        ast::VariantKey::NumberLiteral {
+            value: get_number_literal(ps)?,
+        }
+    } else {
+        ast::VariantKey::Identifier {
+            name: get_identifier(ps)?.name,
+        }
+    };
+
+    ps.skip_blank();
+
+    ps.expect_byte(b']')?;
+
+    Ok(key)
+}
+
+fn get_variants<'p>(ps: &mut ParserStream<'p>) -> Result<Vec<ast::Variant<'p>>> {
+    let mut variants = vec![];
+    let mut has_default = false;
+
+    while ps.is_current_byte(b'*') || ps.is_current_byte(b'[') {
+        let default = ps.take_if(b'*');
+
+        if default {
+            if has_default {
+                return error!(ErrorKind::MultipleDefaultVariants, ps.ptr);
+            } else {
+                has_default = true;
+            }
+        }
+
+        let key = get_variant_key(ps)?;
+
+        let value = get_pattern(ps)?;
+
+        if let Some(value) = value {
+            variants.push(ast::Variant {
+                key,
+                value,
+                default,
+            });
+            ps.skip_blank();
+        } else {
+            return error!(ErrorKind::MissingValue, ps.ptr);
+        }
+    }
+
+    if !has_default {
+        error!(ErrorKind::MissingDefaultVariant, ps.ptr)
+    } else {
+        Ok(variants)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum TextElementTermination {
```

Comment:
All the enums here need docs, which explain the _why_'s of this code. I suggested how to improve the docs of `CarriageReturn` in a comment below.

Reply:
Can `TextElementTermination` be combined with `TextElementPosition`?

Reply:
Sure.